Suppress fallback and ambiguity errors

nikomatsakis opened the pull request:
If the infcx has observed other errors, then suppress both default type
parameter fallback (which can be unreliable, as the full constraint set
is not available) and errors related to unresovled
variables (annoyingly, integer type variables cannot currently be
unified with error, so that has to be a separate mechanism). Also add a
flag to `infcx` to allow us to independently indicate when we have
observed an error and hence should trigger this suppression mode.

Fixes #31997 

cc @alexcrichton 
r? @arielb1

arielb1 commented on src/librustc/middle/infer/sub.rs:
```diff
@@ -86,6 +86,7 @@ impl<'a, 'tcx> TypeRelation<'a, 'tcx> for Sub<'a, 'tcx> {
             }
 
             (&ty::TyError, _) | (_, &ty::TyError) => {
+                infcx.set_tainted_by_errors();
```
When should this be hit?

nikomatsakis replied:
@arielb1 

> "causing damage"

Yes, we could remove it. I don't think it's needed -- rather, we should set the flag at the *source* of the errors, as I later did. (I could also assert when setting the flag that we are not in a snapshot, just to be safe.)

arielb1 commented on src/librustc_typeck/check/cast.rs:
```diff
@@ -121,8 +121,27 @@ impl<'tcx> CastCheck<'tcx> {
+        // modified to push all unresolved inference variables to
+        // ty-err, but it's STILL possible to see fallback for
+        // integral/float variables, because those cannot be unified
+        // with ty-error.
```
Casts *can* resolve types. I would rather change the `type_is_known_to_be_sized` path (both here and in the other place) to be less violent.

arielb1 commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -1771,16 +1775,37 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        // encountered type-checking errors. Therefore, if we think we saw
+        // some errors in this function, just resolve all uninstanted type
+        // varibles to TyError.
+        if self.infcx().is_tainted_by_errors() {
```
If you are doing your best attempt to break out of inference, why not *just do it*?

nikomatsakis replied:
@arielb1 

> If you are doing your best attempt to break out of inference, why not just do it?

You mean why unify with error? The idea was to taint these variables in a way that would cause other errors to be suppressed if they should refer to them. Annoyingly, it didn't entirely work because of integral variables, so I still needed a few tests elsewhere.

alexcrichton commented:
:mahjong:

nikomatsakis commented:
@arielb1 

> Casts can resolve types.

Do you see this as a problem for this patch? Can you elaborate?

> I would rather change the type_is_known_to_be_sized path (both here and in the other place) to be less violent.

Hmm, ok. What would you like it to do? Just assume true?

> Why are integer variables a problem here (maybe we can change ty::cast to be more allowing?).

Because I can't unify them with ty-error.

nikomatsakis commented:
@arielb1 I'm not really sure where you are in your review here. Do you think I should change something? If so, what you would you prefer? (Or if you're just busy, that's fine too.)

arielb1 commented:
@nikomatsakis 

I would prefer to just treat `type_is_known_to_be_sized` as true if we are checking whether `_: Sized` and there are errors (for example, `() as *const _` should still be an error) and to have `ty::cast` treat integer/float variables as actual integer/floats.

> If you are doing your best attempt to break out of inference, why not *just do it*?

I was working on a patch that didn't do `select_all_obligations_or_error` but rather tried to report more specific errors if there were unresolved type variables. That's it, before I became somewhat busy.

Isn't cast checking the only thing downstream of default obligation? And I think we can solve it by being more careful with `type_is_known_to_be_sized`. And then we can figure out how much undefined-variable-error-reporting we want to do.

arielb1 commented:
I was also thinking about propagating `TyErr`, at least in method selection and for closure arguments to invalid methods.

nikomatsakis commented:
@arielb1 so I removed that error suppression logic out of the casts. It didn't seem to cause any new spurious errors. I decided against modifying the logic for when a type is considered sized since it didn't seem necessary. I've also gotten make check passing and rebased. Take another look and tell me what you think. Given the changes to the errors (lots of duplicates removed) this seems like a positive step to me, and I *think* it no longer has the invasive quality you were concerned about.

arielb1 commented:
Does this fix `let x: *const u32 = 0 as _;`? If so, add an rpass test.

arielb1 commented:
Eh, it shouldn't. Maybe put that fix in a separate PR instead?

arielb1 commented on src/librustc_typeck/check/cast.rs:
```diff
@@ -111,17 +113,34 @@ enum CastError {
+        // For better error messages, we try to check whether the
+        // target type is known to be sized now (we will also check
+        // later, once inference is more complete done).
+        if !fcx.type_is_known_to_be_sized(cast_ty, span) {
```
What is *constructor* doing randomly emitting error messages?

BTW, why do we need the early error message? It causes compilation errors in some cases.

nikomatsakis replied:
@arielb1 

> What is constructor doing randomly emitting error messages?

As you well know, a constructor is just another function... seems like as good a time as any. And the return type (in particular, the `Result` with `ErrorReported`) clearly indicates that an error is reported. 

My thinking was basically that instantiating a cast check does some checks and returns, potentially, further checks that may need to be done later. I'm happy to rename.

> BTW, why do we need the early error message? It causes compilation errors in some cases.

Yes, this doesn't seem optimal, I agree. I was just preserving the existing behavior -- but it seems like it could go wrong, if only in unusual situations like `x as _` or `x as Bar<_>`

I originally moved this check to be done at the end, but then you get derived errors later. For example:

    let x = foo as SomeTrait;

will give an error because the type of `x` is `SomeTrait`, which is not `Sized`. (Actually, it then later gives a error about "cast to unsized type".)

Probably the best thing would be to rewrite `type_is_known_to_be_sized` to return "yes, no, maybe" and only error out on *no*. (Or, for simplicity, just check specifically for casts to traits or slices, and leave the more general function for the end.)

nikomatsakis commented:
@arielb1 pushed a new commit that tweaks the approach. Unfortunately, this doesn't seem to be quite enough to make that test work. I didn't dig into why, but you get a "type must be known" error.

nikomatsakis commented:
oh, I see, because it calls `structurally_resolved_type`

nikomatsakis commented:
OK, @arielb1, pushed a new version that fixes the case you mentioned.

nikomatsakis commented:
r? @pnkfelix (@arielb1 seems busy)

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 89bbd2c has been approved by `arielb1`

<!-- @bors r=arielb1 89bbd2c8b75f31f06496158e8e309557cfeaeed5 -->

bors commented:
:hourglass: Testing commit 89bbd2c with merge 4937dda...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3097)

alexcrichton commented:
@bors: retry

On Thu, Apr 21, 2016 at 4:17 PM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-msvc-32-opt
> <http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3097>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/32258#issuecomment-213155541>
>

bors commented:
:zap: Previous build results for [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/175), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/798), [auto-linux-64-x-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armsf/builds/388), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2028), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/987), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/801), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/268) are reusable. Rebuilding only [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8853), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8888), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2120), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8827), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8815), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/392), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8810), [auto-linux-64-x-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armhf/builds/397), [auto-linux-64-x-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-freebsd/builds/400), [auto-linux-64-x-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-netbsd/builds/389), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3969), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8859), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8870), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8831), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/793), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3889), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3905), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3900), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3879), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3097), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/179), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3891), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/393), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/810)...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/395)

arielb1 commented:
@bors retry

bors commented:
:hourglass: Testing commit 89bbd2c with merge 3247f37...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3886)

arielb1 commented:
@bors retry

What the **** is going with our windows bots?

bors commented:
:zap: Previous build results for [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/180), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/803), [auto-linux-64-x-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armsf/builds/393), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2033), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3975), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/992), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/816) are reusable. Rebuilding only [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8859), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8894), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2126), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8833), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8821), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/399), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8816), [auto-linux-64-x-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armhf/builds/403), [auto-linux-64-x-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-freebsd/builds/406), [auto-linux-64-x-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-netbsd/builds/395), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8866), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8877), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8837), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/799), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3897), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3912), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/806), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3908), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3886), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/273), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3105), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/185), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3898), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/401)...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3924)

alexcrichton commented:
@bors: retry

On Fri, Apr 22, 2016 at 8:46 AM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-gnu-32-opt
> <http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3924>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/32258#issuecomment-213482295>
>

bors commented:
:hourglass: Testing commit 89bbd2c with merge d43fb29...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3122)

alexcrichton commented:
@bors: retry

On Fri, Apr 22, 2016 at 5:54 PM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-msvc-32-opt
> <http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3122>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/32258#issuecomment-213640156>
>

bors commented:
:hourglass: Testing commit 89bbd2c with merge b428b62...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3124)

alexcrichton commented:
@bors: retry

On Fri, Apr 22, 2016 at 9:47 PM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-msvc-32-opt
> <http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3124>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/32258#issuecomment-213663352>
>

bors commented:
:hourglass: Testing commit 89bbd2c with merge 3a44be5...

bors commented:
:broken_heart: Test failed - [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8898)

nikomatsakis commented:
@bors r=arielb1

bors commented:
:pushpin: Commit b3d54a2 has been approved by `arielb1`

<!-- @bors r=arielb1 b3d54a2c0e793768cd9ce4a7887a75e3ec356cb5 -->

bors commented:
:hourglass: Testing commit b3d54a2 with merge 90318b8...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8915), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8950), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/42), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/234), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/42), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/43), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/43), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/43), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2182), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8889), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8877), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/455), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/857), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8872), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2087), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4030), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8922), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8933), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8893), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/855), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/43), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3954), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3969), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/862), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3965), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3943), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/328), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3161), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/240), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3955), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/458), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/870)

nagisa commented:
Tagging this as relnotes because of https://github.com/rust-lang/rust/issues/33657
